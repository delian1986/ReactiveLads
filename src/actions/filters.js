import { SELECT_MATERIAL_TYPE, SELECT_COLOR, SELECT_TAG } from "./constants";

export const selectMaterialType = (material) => {
  return {
    type: SELECT_MATERIAL_TYPE,
    payload: material
  };
};

export const selectColor = (color) => {
  return {
    type: SELECT_COLOR,
    payload: color
  };
};

export const selectTag = (tag) => {
  return {
    type: SELECT_TAG,
    payload: tag
  };
};
