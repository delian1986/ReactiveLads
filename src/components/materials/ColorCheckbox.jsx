import { Component } from "react";
import PropTypes from "prop-types";

export class ColorCheckbox extends Component {
  constructor() {
    super();

    this.state = {
      outline: false
    };

    this.handleChange = this.handleChange.bind(this);
  }

  handleChange() {
    this.setState({
      outline: !this.state.outline
    });
    this.props.selectColor(this.props.id);
  }

  render() {
    return (
      <div
        style={{
          background: this.props.hex,
          height: "25px",
          width: "25px",
          position: "relative",
          float: "left",
          margin: "5px",
          outline: this.state.outline ? "auto" : "none"
        }}
        onClick={this.handleChange}
      />
    );
  }
}

ColorCheckbox.propTypes = {
  id: PropTypes.number.isRequired,
  hex: PropTypes.string.isRequired,
  selectColor: PropTypes.func.isRequired
};
