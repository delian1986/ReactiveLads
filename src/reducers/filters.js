import {
  SELECT_MATERIAL_TYPE,
  SELECT_COLOR,
  SELECT_TAG
} from "../actions/constants";

const initialState = {
  selectedMaterialTypes: [],
  selectedColors: [],
  selectedTags: []
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SELECT_MATERIAL_TYPE:
      if (state.selectedMaterialTypes.includes(action.payload)) {
        return {
          ...state,
          selectedMaterialTypes: state.selectedMaterialTypes.filter(
            (m) => m !== action.payload
          )
        };
      } else {
        return {
          ...state,
          selectedMaterialTypes: [...state.selectedMaterialTypes, action.payload]
        };
      }
    case SELECT_COLOR:
      if (state.selectedColors.includes(action.payload)) {
        return {
          ...state,
          selectedColors: state.selectedColors.filter((c) => c !== action.payload)
        };
      } else {
        return {
          ...state,
          selectedColors: [...state.selectedColors, action.payload]
        };
      }
    case SELECT_TAG:
      if (state.selectedTags.includes(action.payload)) {
        return {
          ...state,
          selectedTags: state.selectedTags.filter((c) => c !== action.payload)
        };
      } else {
        return {
          ...state,
          selectedTags: [...state.selectedTags, action.payload]
        };
      }
    default:
      return state;
  }
};
