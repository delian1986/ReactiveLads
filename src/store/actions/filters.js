import {
  SELECT_MATERIAL_TYPE,
  SELECT_COLOR,
  SELECT_TAG,
  IS_IN_FAVORITES_MODE,
  SET_ALL_FILTERS,
  SET_SEARCH_QUERY,
  RESET_ALL_FILTERS
} from "./constants";

export const selectMaterialType = (material) => {
  return {
    type: SELECT_MATERIAL_TYPE,
    payload: material
  };
};

export const selectColor = (color) => {
  return {
    type: SELECT_COLOR,
    payload: color
  };
};

export const selectTag = (tag) => {
  return {
    type: SELECT_TAG,
    payload: tag
  };
};

export const setAllFilters = (presets) => {
  return {
    type: SET_ALL_FILTERS,
    payload: presets
  };
};

export const toggleFavoritesMode = (mode) => {
  return {
    type: IS_IN_FAVORITES_MODE,
    payload: mode
  };
};

export const setSearchQuery = (value) => ({
  type: SET_SEARCH_QUERY,
  payload: value
});

export const resetAllFilters = () => ({
  type: RESET_ALL_FILTERS
});
