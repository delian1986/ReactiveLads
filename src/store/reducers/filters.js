import {
  SELECT_MATERIAL_TYPE,
  SELECT_COLOR,
  SELECT_TAG,
  IS_IN_FAVORITES_MODE,
  SET_ALL_FILTERS,
  SET_SEARCH_QUERY,
  RESET_ALL_FILTERS
} from "../actions/constants";

const initialState = {
  selectedMaterialTypes: [],
  selectedColors: [],
  selectedTags: [],
  searchQuery: "",
  isInFavoritesMode: false
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SELECT_MATERIAL_TYPE:
      if (state.selectedMaterialTypes.includes(action.payload)) {
        return {
          ...state,
          selectedMaterialTypes: state.selectedMaterialTypes.filter(
            (m) => m !== action.payload
          )
        };
      } else {
        return {
          ...state,
          selectedMaterialTypes: [...state.selectedMaterialTypes, action.payload]
        };
      }
    case SELECT_COLOR:
      if (state.selectedColors.includes(action.payload)) {
        return {
          ...state,
          selectedColors: state.selectedColors.filter((c) => c !== action.payload)
        };
      } else {
        return {
          ...state,
          selectedColors: [...state.selectedColors, action.payload]
        };
      }
    case SELECT_TAG:
      if (state.selectedTags.includes(action.payload)) {
        return {
          ...state,
          selectedTags: state.selectedTags.filter((c) => c !== action.payload)
        };
      } else {
        return {
          ...state,
          selectedTags: [...state.selectedTags, action.payload]
        };
      }
    case SET_ALL_FILTERS:
      return {
        ...state,
        selectedTags: action.payload.tags,
        selectedColors: action.payload.colors,
        selectedMaterialTypes: action.payload.types,
        searchQuery: action.payload.searchQuery,
        isInFavoritesMode: action.payload.favoritesMode
      };
    case SET_SEARCH_QUERY:
      return {
        ...state,
        searchQuery: action.payload
      };
    case RESET_ALL_FILTERS:
      return initialState;
    case IS_IN_FAVORITES_MODE:
      return {
        ...state,
        isInFavoritesMode: action.payload
      };
    default:
      return state;
  }
};
