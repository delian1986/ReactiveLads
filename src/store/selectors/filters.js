import { createSelector } from "reselect";

const isInFavMode = (state) => state.filters.isInFavoritesMode;

export const getIsInFavoritesMode = createSelector(
  isInFavMode,
  (favMode) => favMode
);
