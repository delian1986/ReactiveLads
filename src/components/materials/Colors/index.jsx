import { Component } from "react";
import { ColorCheckbox } from "../ColorCheckbox";
import PropTypes from "prop-types";

export class Colors extends Component {
  componentDidMount() {
    this.props.fetchColors();
  }

  render() {
    const { colors, selectColor } = this.props;
    return (
      <>
        <h5>Colors</h5>
        {colors.length > 0 ? (
          <div style={{ display: "flow-root" }}>
            <h5>Colors</h5>
            {colors.map((color) => (
              <ColorCheckbox
                key={color.id}
                id={color.id}
                hex={color.hex}
                selectColor={selectColor}
              />
            ))}
          </div>
        ) : (
          <div className="d-flex justify-content-center">
            <div className="spinner-border" role="status">
              <span className="sr-only">Loading...</span>
            </div>
          </div>
        )}
      </>
    );
  }
}

Colors.propTypes = {
  colors: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      hex: PropTypes.string,
      selectColor: PropTypes.func
    })
  ),
  fetchColors: PropTypes.func.isRequired,
  selectColor: PropTypes.func.isRequired
};
