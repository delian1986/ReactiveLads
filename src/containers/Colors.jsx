import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { Colors } from "../components/materials/Colors";

import { selectColor } from "../actions/filters";

const mapStateToProps = (state) => {
  return {
    colors: state.colors,
    selectedColorsFilter: state.filters.selectedColors
  };
};

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      selectColor: selectColor
    },
    dispatch
  );

export default connect(mapStateToProps, mapDispatchToProps)(Colors);
