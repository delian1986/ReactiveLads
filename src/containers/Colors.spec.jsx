import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import Colors from "./Colors";
import { mount } from "enzyme";
import { ColorCheckbox } from "../components/materials/ColorCheckbox";
import { selectColor } from "../actions/filters";

const mockStore = configureStore([]);

describe("tests <Colors/> component", () => {
  let store;
  let component;
  beforeEach(() => {
    store = mockStore({
      colors: [
        {
          id: 1,
          hex: "data",
          background: "black"
        }
      ],
      filters: {
        selectedColors: [3, 5]
      }
    });
    store.dispatch = jest.fn();
    component = renderer.create(
      <Provider store={store}>
        <Colors component={() => <></>} />
      </Provider>
    );
  });
  describe("Test rendering component", () => {
    it("should render with given state from Redux store", () => {
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  describe("Integration test", () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mount(
        <Provider store={store}>
          <Colors />
        </Provider>
      );
    });
    it("should deselect <ColorCheckbox/>", () => {
      expect(wrapper.find(ColorCheckbox).length).toEqual(1);
      wrapper.find(ColorCheckbox).simulate("click");
      expect(store.dispatch).toHaveBeenCalledWith(selectColor(1));
    });
  });

  describe("Should be deselected", () => {
    let store;
    let wrapper;
    beforeEach(() => {
      store = mockStore({
        colors: [
          {
            id: 1,
            hex: "data",
            background: "black"
          }
        ],
        filters: {
          selectedColors: [3, 5]
        }
      });
      store.dispatch = jest.fn();
      wrapper = mount(
        <Provider store={store}>
          <Colors />
        </Provider>
      );
    });
    it("should render selected", () => {
      expect(wrapper.find(ColorCheckbox).length).toEqual(1);
      expect(wrapper.find(ColorCheckbox).get(0).props.isSelected).toEqual(false);
    });
  });
  describe("Should be selected", () => {
    let store;
    let wrapper;
    beforeEach(() => {
      store = mockStore({
        colors: [
          {
            id: 1,
            hex: "data",
            background: "black"
          }
        ],
        filters: {
          selectedColors: [1, 5]
        }
      });
      store.dispatch = jest.fn();
      wrapper = mount(
        <Provider store={store}>
          <Colors />
        </Provider>
      );
    });
    it("should render selected", () => {
      expect(wrapper.find(ColorCheckbox).length).toEqual(1);
      expect(wrapper.find(ColorCheckbox).get(0).props.isSelected).toEqual(true);
    });
  });
});
