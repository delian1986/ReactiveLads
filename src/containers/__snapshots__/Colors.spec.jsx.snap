// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests <Colors/> component Test rendering component should render with given state from Redux store 1`] = `
Array [
  <h5>
    Colors
  </h5>,
  <div
    style={
      Object {
        "display": "flow-root",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "background": "data",
          "float": "left",
          "height": "25px",
          "margin": "5px",
          "outline": "none",
          "position": "relative",
          "width": "25px",
        }
      }
    />
  </div>,
]
`;
